src/img/img-xpm.c

```diff
@@ -0,0 +1,1968 @@
+	color_ptr->blue = (found->blue * 65535) / 255;
+
+	return 1;
+}
+
+/*
+ *----------------------------------------------------------------------
+ *
+ * parse_color --
+ *
+ *	Partial implementation of X color name parsing interface.
+ *
+ * Results:
+ *	Returns 1 on success.
+ *
+ * Side effects:
+ *	None.
+ *
+ *----------------------------------------------------------------------
+ */
+
+static int parse_color(const char *spec, struct xpm_color *color_ptr)
+{
+	if (spec[0] == '#') {
+		int i;
+		unsigned int red, green, blue;
+
+		i = strlen(spec + 1);
+		if (i % 3) {
+			return 0;
+		}
+		i /= 3;
+
+		if (i == 4) {
+			if (sscanf(spec + 1, "%4x%4x%4x", &red, &green, &blue) != 3) {
+				return 0;
+			}
+			color_ptr->red = red;
+			color_ptr->green = green;
+			color_ptr->blue = blue;
+		} else if (i == 1) {
+			if (sscanf(spec + 1, "%1x%1x%1x", &red, &green, &blue) != 3) {
+				return 0;
+			}
+			color_ptr->red = (red * 65535) / 15;
+			color_ptr->green = (green * 65535) / 15;
+			color_ptr->blue = (blue * 65535) / 15;
+		} else if (i == 2) {
+			if (sscanf(spec + 1, "%2x%2x%2x", &red, &green, &blue) != 3) {
+				return 0;
+			}
+			color_ptr->red = (red * 65535) / 255;
+			color_ptr->green = (green * 65535) / 255;
+			color_ptr->blue = (blue * 65535) / 255;
+		} else /* if (i == 3) */ {
+			if (sscanf(spec + 1, "%3x%3x%3x", &red, &green, &blue) != 3) {
+				return 0;
+			}
+			color_ptr->red = (red * 65535) / 4095;
+			color_ptr->green = (green * 65535) / 4095;
+			color_ptr->blue = (blue * 65535) / 4095;
+		}
+	} else {
+		if (!find_color(spec, color_ptr)) {
+			return 0;
+		}
+	}
+	return 1;
+}
+
+static int xpm_seek_string(FILE *infile, const char *str)
+{
+	char instr[1024];
+
+	while (!feof(infile)) {
+		if (fscanf(infile, "%1023s", instr) < 0) {
+			return 0;
+		}
+		if (strcmp(instr, str) == 0) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+static int xpm_seek_char(FILE *infile, char c)
+{
+	int b, oldb;
+
+	while ((b = getc(infile)) != EOF) {
+		if (c != b && b == '/') {
+			b = getc(infile);
+			if (b == EOF) {
+				return 0;
+			} else if (b == '*') {
+				/* we have a comment */
+				b = -1;
+				do {
+					oldb = b;
+					b = getc(infile);
+					if (b == EOF) {
+						return 0;
+					}
+				} while (!(oldb == '*' && b == '/'));
+			}
+		} else if (c == b) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+static int xpm_read_string(FILE *infile, char **buffer, uint32_t *buffer_size)
+{
+	int c;
+	uint32_t cnt = 0, bufsiz, ret = 0;
+	char *buf;
+
+	buf = *buffer;
+	bufsiz = *buffer_size;
+	if (!buf) {
+		bufsiz = 10 * sizeof(char);
+		buf = (char *)calloc(bufsiz, sizeof(char));
+	}
+
+	do {
+		c = getc(infile);
+	} while (c != EOF && c != '"');
+
+	if (c != '"') {
+		goto out;
+	}
+
+	while ((c = getc(infile)) != EOF) {
+		if (cnt == bufsiz) {
+			uint32_t new_size = bufsiz * 2;
+
+			if (new_size > bufsiz) {
+				bufsiz = new_size;
+			} else {
+				goto out;
+			}
+
+			buf = (char *)xrealloc(buf, bufsiz);
```

Comment:
Using `struct buf` makes this a lot easier. If you like, there is a more cleaned-up version of this parser in #2193. I didn't see your PR first, sorry...

Reply:
Agree with using `struct buf`.